astropy/io/fits/column.py

```diff
@@ -1389,19 +1389,21 @@ def __getattr__(self, name):
 
         Implements for example self.units, self.formats, etc.
         """
-
         cname = name[:-1]
         if cname in KEYWORD_ATTRIBUTES and name[-1] == 's':
             attr = []
-            for col in self:
+            for col in self.columns:
                 val = getattr(col, cname)
-                if val is not None:
-                    attr.append(val)
-                else:
-                    attr.append('')
+                attr.append(val if val is not None else '')
             return attr
         raise AttributeError(name)
 
+    def _update_attrs(self, name=None):
+        """Get or update local lists of column attributes"""
+        attrs = [name] if name is not None else KEYWORD_ATTRIBUTES
+        for name in attrs:
+            setattr(self, name + 's', self.__getattr__(name + 's'))
```

Comment:
Ah, I think I understand now what you're doing.  

My one problem with this is that it's making already far too fragile code (most of which I've long wanted to get rid of entirely!) even more fragile.  A forgetting `_update_attrs` in subtle places could lead to big problems, and it's not at all clear to the casual observer what this is even for or when/why it needs to be called.

It also means that there's a second copy of all the column attributes, and I'm really trying to avoid that, by keeping all information in one place.  The `ColDefs` class is definitely a mess, and is something of a historical artifact (it shouldn't even be returning lists for the `<attr>s` attributes, but rather tuples, and it's questionable even whether those attributes are useful at all).

I think the rest of this PR, but I'm tempted to do away with this, and instead make less use of the `ColDefs.<attr>s` attributes where possible (e.g. instead of `cols.names[i]` use `cols[i].name`, etc.)